### PR TITLE
Migrate to use the Fixie Voice SDK.

### PR DESCRIPTION
**Pull request by mdwelsh**
This PR updates app/agent/page.tsx to use the Fixie VoiceSession API instead of the hand-rolled stuff currently in there.

To make this work, I had to adjust the initialization logic to more closely match what we do in HiSanta, since we've changed how the startup is done.

I made the start/stop button the only way to toggle the states, rather than tapping anywhere, just to simplify the code a bit.

Happy to change this up. Thinking it might make sense to do some tweaks to the UI so it's a more useful debugging tool without needing keyboard shortcuts and URL path editing.

### CODE REVIEW

File: src/app/agent/page.tsx [hunk 1]
```diff
@@ -270,133 +283,134 @@ const AgentPageComponent: React.FC = () => {
...
-      asrProvider,
-      asrModel,
-      asrLanguage,
-      ttsProvider,
-      ttsModel,
-      ttsVoice,
-      model,
+
+  // Returns true if the voice session is active.
+  const active = useCallback(() => {
```
Comment (juberti) [inline src/app/agent/page.tsx:299]: I don't fully understand the `useCallback` usages - my understanding is that this hook is entirely a perf optimization, and as these functions are pretty minimal it's not clear why the hook is needed. Generally, unless there's some new incompatibility, I'd prefer to lift-and-shift the current code (modulo any renames) and then rework the React handling separately if needed.

File: src/app/agent/page.tsx [hunk 2]
```diff
@@ -270,133 +283,134 @@ const AgentPageComponent: React.FC = () => {
...
-    }
-  };
+  }, [agentId]);
+
+
   // Spacebar starts or interrupts. Esc quits.
   // C toggles the chooser. S toggles the stats.
-  const onKeyDown = (event: KeyboardEvent) => {
+  const onKeyDown = useCallback((event: KeyboardEvent) => {
+  // Either interrupt the vcoice session, or start it.
```
Comment (juberti) [inline src/app/agent/page.tsx]:
```suggestion
  // Either interrupt the voice session, or start it.
```

File: src/app/agent/page.tsx (hunk 1 above)
Comment (mdwelsh) [reply src/app/agent/page.tsx:299]: That was my original plan, but I was getting all kinds of lint errors with the existing code. The issue has to do with how these functions were called by useEffect hooks. This leads to a transitive dependency on the dependencies of the functions (since useEffect needs to declare dependencies unless you are calling the effect unconditionally on every render, which can be expensive). The useCallback allows us to memoize the callbacks unless their own dependencies change, and then, by extension, the useEffect hooks don't need to be rerun unless those callbacks have been changed.

The existing code was not doing this correctly and hence was breaking (for me, anyway).

Comment (mdwelsh) [reply src/app/agent/page.tsx:299]: Here's a concrete example of what happens if I don't use `useCallback` for the `active` function (this applies to the other cases too).

* If I make the `active` function a regular function and not a callback, I get the error that this causes the `onKeyDown` callback to get recomputed on every render (since that function calls `active`):

```The 'active' function makes the dependencies of useCallback Hook (at line 438) change on every render. To fix this, wrap the definition of 'active' in its own useCallback() Hook.eslint[react-hooks/exhaustive-deps](https://github.com/facebook/react/issues/14920)```

* To fix this, I can have `onKeyDown` also be a function, and no longer a callback, which means it gets recomputed on every render. This causes the linter to complain that the `useEffect` hook on line 440 (`"Install our handlers, and clean them up on unmount."`) will be run on every render, since `onKeyDown` will change on every render.

* I can take out the dependency array for that `useEffect`, but that - you guessed it - causes it to be run on every render.

With the `useCallback` hooks, the `useEffect` on line 440 is only run whenever the `onKeyDown` callback changes, which would only be run if `active` (and its other dependencies) change.

The existing code was only partially declaring its dependencies, so it was not actually correct in that an effect might not run when it should have.

Basically, the general rule is this: if we're gonna have any `useEffect` hooks with functions as dependencies, those dependencies need to be callbacks themselves, with their own dependencies specified.

Comment (juberti) [reply src/app/agent/page.tsx:299]: OK, thanks for explaining. I'll take another look with this in mind.

File: src/app/agent/page.tsx [hunk 3]
```diff
@@ -270,133 +283,134 @@ const AgentPageComponent: React.FC = () => {
...
-  };
+  }, [agentId]);
+
+
   // Spacebar starts or interrupts. Esc quits.
   // C toggles the chooser. S toggles the stats.
-  const onKeyDown = (event: KeyboardEvent) => {
+  const onKeyDown = useCallback((event: KeyboardEvent) => {
+  // Either interrupt the voice session, or start it.
+  const speak = () => (active() ? voiceSession!.interrupt() : handleStart());
```
Comment (juberti) [inline src/app/agent/page.tsx]: indent looks off here

File: src/app/agent/page.tsx [hunk 4]
```diff
@@ -270,133 +283,134 @@ const AgentPageComponent: React.FC = () => {
...
+    console.log(`init - warming up voice session ${session}`);
+    session.warmup();
+    setVoiceSession(session);
+    return session;
+  }, [agentId, asrProvider, model, ttsProvider, ttsVoice, webrtcUrl]);
+
+  // Called when the main button is clicked.
+  const onButtonClick = useCallback(() => {
+    if (!started) {
+      const session = createSession();
```
Comment (juberti) [inline src/app/agent/page.tsx:383]: the way the session is passed around here seems kind of weird (ie not as an argument to handleStart), and it seems the session is started in multiple places. I also think we want to create and warm up the session on pageload.

File: src/app/agent/page.tsx (hunk 3 above)
Comment (mdwelsh) [reply src/app/agent/page.tsx]: Fixed

File: src/app/agent/page.tsx (hunk 4 above)
Comment (mdwelsh) [reply src/app/agent/page.tsx:383]: So, I'm focused right now on *only* getting the app to work with the Voice SDK, but I realize there are a few ways the initialization logic and such could be improved. You'll recall I spent hours beating my head against this in the HiSanta app -- so, I'm just doing something a little crude that works for now. What I ended up doing in HiSanta was more complex (in part because of the ringtone support, but also because of the same stuff you're noticing here), and in keeping with "lift and shift" as much as possible I didn't want to go rewriting this component from scratch to make it work the same way as HiSanta.

I'm happy to go do that and clean all this stuff up, however, it's gonna take some time and I was under the impression we just wanted a quick-and-dirty port so we could move onto the Voice SDK as a starting point.

I agree that we want to be able to create and warm up the session on pageload. Unfortunately that is hard to do with the way React useEffect works, and will require more rework. Happy to do that in follow PRs.

Comment (juberti) [reply src/app/agent/page.tsx:383]: The thing is that the existing code does do the pageload warmup, and seems to work fine. If we can't preserve that in the updated code, can we just lift-and-shift the existing HTML/JS code then?

**Comment by mdwelsh** [comment]
Justin is going to do this another way
